README
THIS IS RELEASE EARLY CODE!  Maybe that the latest version introduces bugs.

ptybuffer
=========

For the impatient
-----------------

```bash
git clone https://github.com/hilbix/ptybuffer.git
make
sudo make install
```

```bash
mkdir -p ~/bin ~/autostart ~/autorun
cp scripts/autostart.sh ~/bin/
cp scripts/helloworld.sh ~/autostart/
cp autorun/helloworld.sh ~/autorun/
~/bin/autostart.sh
socat - unix:/var/tmp/autostart/$LOGNAME/helloworld.sock
{ crontab -l; echo '* * * * * bin/autostart.sh >/dev/null'; } | crontab -
```

Infos:
```bash
./ptybuffer -h
head test/test.sh
```

About
-----

ptybuffer

- daemonizes.  With option -d it runs in foreground

- calls a process (arg2 and above) with a pty such that stdlib
  usually makes output line buffered.

- gathers the output history of the pty for later view.  Currently
  this is 1000 reads (not neccessarily lines, see option -n).

- waits for connects to the unix socket (command line argument 1)

- outputs the pty history to the connected unix socket and
  continuously sends all new output to the connected unix socket.

- reads (full) lines from the connected unix sockets and send them to
  the pty (line can be max. 4095 bytes long).

- allows multiple concurrent connects to the unix socket.  This way
  several persons in parallel can control one process.

This is better than running background tasks using screen or tmux:

- Screen blocks all connects if one session freezes.  ptybuffer is
  thought to always let the program run, regardless what mess the
  socket connections might do.  In case of a connect stall it just
  discards data to the frozen connect and if the connect catches up
  it tells it the fact.

- Concurrent access is no problem.  Lines are always send as full lines.

- History of output is clearly defined, not just something like a
  window.

- Naturally scrolling logfile support.  Eases development in case you
  have excessive debugging active and want to see the output as soon
  as it comes from the program, even without shell access.

- You can control where the socket is.  This improves oversight in
  case you have running several commands in parallel.

- Easy to use from web scripting languages like PHP which can directly
  connect to a Unix Domain Socket.  Really no need to use error prone
  constructs like expect.

- If the central process of screen or tmux dies, all background tasks
  die as well.  Ptybuffer runs independently for each task.

- Ptybuffer does not interpret the terminal.  So no complicated code
  which might hide bugs.


Bugs:
=====

- ptybuffer does not unlink() the socket if option -f and -c are
  missing.  This is a feature.

- The program running under pty control is not considered to stop.  If
  this ever happens, ptybuffer just exits without making sure that all
  data is sent to connected sockets.  You can repair this by wrapping
  into a script which, for example, ends in a read with some timeout.

- New version have not been tested much before release.

- If you send data to ptybuffer and close the socket very fast, the
  EOF on the write side closes the socket before everything can be
  read from the read side by the select-loop.  Luckily, usually not
  much is sent to the controlled process this way (one line or so), so
  a short sleep should be enough for ptybuffer to gather the data.  To
  fix this BUG I must implement the socket flags first and restructure
  the the EOF processing a bit - no time for it now, it currently
  works good enough for me.


Example:
========

See introductory comment in test/test.sh

See scripts/autostart.sh with example scripts/helloworld.sh

For a PHP usage example see the php/ subdirectory.

You need a program capable to send data to unix sockets to control
the program running in ptybuffer.  `socat` is your friend.


DISCLAIMER:
===========

USE AT YOUR OWN RISK!  I CANNOT ACCEPT ANY LIABLILITY FOR ANYTHING!
THIS *IS* RELEASE EARLY CODE!  IT IS CONSIDERED TO BE INSTABLE!

DON'T TRUST SOURCES!  READ THEM!  READ THEM AGAIN!  AND BE SURE THERE
ARE NO KNOWN BACKDOORS IN THEM!  IT'S OPEN SOURCE, SO YOU CAN CHECK!

I tried my best.  However I am human.  So I make mistakes.  Be prepared.
All I guarantee is, that I never do anything to harm *you* by purpose.

Copyright (C)2004-2014 by Valentin Hilbig

ptybuffer may be distributed freely under the conditions of the
GPL2 (GNU General Public License version 2) or higher.

-Tino


ChangeLog
2016-10-11  Valentin Hilbig  <tino@medusa.03.softkill.org>

	* ptybuffer.c (daemonloop): Correcting end loop condition, now
	ptybuffer terminates when there are no more sockets.
	(sock_process): correcting close condition.  Close main socket if
	there are no other sockets.  This means, as long as some socket
	hangs, you can open new sockets.
	(struct ptybuffer_params, struct ptybuffer, main, daemonloop): new
	option -w to keep main socket open until last socket disconnects.
	(sock_process): close main socket as soon as pty is closed, unless
	option -w is given.

2016-08-28  Valentin Hilbig  <tino@medusa.03.softkill.org>

	* ptybuffer.c (main): Bugfix: Stdin of the child was accidentally
	closed when using -d.  This changes the child interface a bit,
	so minor version needs an increment (even there is no new feature).
	(daemonloop): moved socket creation to correctly show socket count
	(connect_process): do not count socket which is about to be closed
	(connect_process): write last output on PTY close
	(master_process): write all input on master socket close
	(daemonloop): do not terminate as long as there is data to write
	(main): stdin/stdout socket wrapper for proper operation

2007-09-21    <tino@firebird.03.softkill.net>

	* ptybuffer.c (do_check): resource leakage fixed: Option -c forgot
	to close the socket.

2007-09-18    <tino@firebird.03.softkill.net>

	* ptybufferconnect.c (ignore_fn): added, should work now as expected

2007-09-06    <tino@firebird.03.softkill.net>

	* script/autostart.sh: "start"-message now is send to STDERR too,
	so you can redirect STDOUT to /dev/null to suppress the "ok"
	messages to cron.

2007-08-29    <tino@firebird.03.softkill.net>

	* ptybufferconnect.c (out_fn): bugfix: Send to stdout, not stdin

	* ptybuffer.c (connect_process): found where the buggy lines from
	ptybuffer_new_fd must be put.  And fixed the bug in main where the
	wrong variable length (int instead of long) was passed to getopt
	(thus the variable was not initialized correctly).
	(main): option -k added and option -i implemented
	(connect_process): option -i and -k implemented
	(send_to_pty): added (code moved from connect_process)

	* TODO (BUGs): added

	* ptybufferconnect.c (main): option -i added

	* ptybuffer.c (connect_process): bugfix: min was int, not long
	long as needed.  Long running ptybuffers (more than 2 billion IO
	operations) thus were not able to output any data to the socket
	and eat up 100% CPU.
	(ptybuffer_new_fd): I am really not sure what I wanted to do with
	this lines I now disabled.  buf->screenpos must be initialized to
	0 to print the history buffer overrun lines correctly.  However it
	was initizlized to a negative value by these lines if there were
	less than history_tail lines in the buffer.  Perhaps an overrun
	issue or forgotten buggy code before the skipped infoblocks were
	printed?  Problem showed up on 64 bit platforms first, I am really
	puzzled why not before.

2007-08-25    <tino@firebird.03.softkill.net>

	* Makefile.tino (DISTCLEAN): cleans for php/ test socket

2007-08-24    <tino@firebird.03.softkill.net>

	* php/.htaccess, php/sock/.htaccess: added

	* php/index.html, php/inp.php, php/out.php, php/startshell.sh:
	added

	* test/stress.sh: Option -p and -o added for tests and names of
	intermediate files changed

	* ptybuffer.c (main): timstamp option added (actually it was
	already there, but the option was missing)
	(file_timestamp): logging changed, on option -o only prefix with
	date and time, not a ':'

2007-06-01    <tino@firebird.03.softkill.net>

	* ptybuffer.c (main): outfile - and logfile - now do work again.
	(main): SIGPIPE now is ignored globally, lools like the previous
	change (form 2007-04-29) was complete crap ..
	(file_open): logfile and outfile are no more written if no
	filename is given on commandline.  As - works again, you
	log/output to stderr/stdout again.
	(file_log): Now has no sideffects on errno.  Errno must not be
	changed by the (new) error logging of the accept() call.

2007-04-29    <tino@firebird.03.softkill.net>

	* ptybuffer.c (parent): SIGPIPE blocking moved here (I do not
	remember why I needed this, perhaps it was just to be sure).
	(main): When the child returns, SIGCHLD is delivered to the
	process immediately.

2007-04-19    <tino@firebird.03.softkill.net>

	* ptybuffer.c (file_out): removed deprecated cast of lvalue (there
	are things which are hard to explain if you don't know what this
	is).

2007-04-07    <tino@firebird.03.softkill.net>

	* script/autostart.sh: Typo corrected (startd -> started)
	* script/autostart.sh (PATH): added in case it's used from cron

2007-03-31    <tino@firebird.03.softkill.net>

	* script/autostart.sh (for): Forgot to change echo into log

2007-03-26    <tino@firebird.03.softkill.net>

	* script/autostart.sh (log): added better logging
	(RUNDIR): in Shell && does not bind tighter than ||, fixed

	* ptybuffer.c (main, file_timestamp): pid reset to show correct
	PID after forks
	(file_timestamp): output corrrected to what it ought to be
	(main): logfile paths are now logged (in case realpath fails)

2007-02-26    <tino@firebird.03.softkill.net>

	* ptybuffer.c (file_open, file_flush_close, file_timestamp): added
	(file_out): timestamping added

2006-12-02    <tino@firebird.03.softkill.net>

	* ptybuffer.c (master_process, log_childstatus): logging made more
	clear

2006-11-11    <tino@firebird.03.softkill.net>

	* ptybuffer.c (main): realpath added to logfile and outfile.

2006-08-11    <tino@firebird.03.softkill.net>

	* ptybuffer.c: dirty.h removed, so changed all occurences of
	"dirty" names to the full name.
	(main): renamed HISTORY_LENGTH to PTYBUFFER_HISTORY_LENGTH
	(main): typo corrected which always gave "missing command" if
	option -c was not present.
	(daemonloop): now return void

2006-07-26    <tino@firebird.03.softkill.net>

	* ptybuffer.c (ptybuffer_new_fd, main): history_tail added (option -t)
	(main): options check (-c) and force (-f) added


php/inp.php
<?
# $Header$
#
# Ptybuffer web based shell - input window
#
# This Works is placed under the terms of the Copyright Less License,
# see file COPYRIGHT.CLL.  USE AT OWN RISK, ABSOLUTELY NO WARRANTY.
#
# $Log$
# Revision 1.2  2007-09-21 11:10:15  tino
# CLL notice now in scripts
#
# Revision 1.1  2007/08/24 17:32:21  tino
# first version

$autoscroll=$_REQUEST["autoscroll"];
$action=$_REQUEST["action"];
$confirm=$_REQUEST["confirm"];
$shortcut=$_REQUEST["shortcut"];
$direct=$_REQUEST["direct"];
$port=$_REQUEST["port"];
$ok=$_REQUEST["ok"];
$send=$_REQUEST["send"];

if ($autoscroll=="" && !$_REQUEST["inited"]) $autoscroll=1;
if ($port=="") $port="shell";
if (!preg_match("/^[-.a-zA-Z0-9]*$/",$port))
  $port	= "";

function checked($x)
{
  if ($x!="")
    echo "CHECKED";
}
?>
<HTML>
<BODY BGCOLOR=#ccccff TEXT=#000000 LINK=#880088 VLINK=#0000ff ALINK=#ff0000
 marginwidth=0 marginheight=0 topmargin=0 leftmargin=0 onload=init()>
<SCRIPT><!--
var nextevent, scrolltime;

function scroll()
{
  nextevent=500;
  if (parent["out"])
    {
      if (parent["out"].halted)
        {
	  if (confirm("output window disconnected\nreload window? (no stops scroller)"))
            {
              parent["out"].location.reload();
	      nextevent = 2500;
            }
          else
            {
  	      document.frm.autoscroll.checked = 0;
            }
        }
      else
	{
	  parent["out"].scrollBy(0, 5000);
	}
    }
}

function scroller()
{
  if (!scrolltime && document.frm.autoscroll.checked)
    scrolltime = window.setTimeout("doscroll()", nextevent);
}

function doscroll()
{
  scrolltime = 0;
  scroll();
  scroller();
}

function checkscroll()
{
  if (!scrolltime)
    doscroll();
}

function scrollit()
{
  if (document.frm.autoscroll.checked)
    checkscroll();
  return true;
}

function checkcr(arg)
{
  if ((window.event && window.event.keyCode==13) || (arg && arg.which==13))
    document.frm.submit();
  return true;
}

function init()
{
  scrollit();
  if (document.frm.send)
    {
      document.frm.send.focus();
      document.frm.send.select();
    }
}
//-->
</SCRIPT>
<FORM ACTION="inp.php" NAME=frm METHOD=post>
<INPUT TYPE=hidden NAME=inited VALUE=1>
<TABLE CELLSPACING=0 CELLPADDING=0><TR><TD VALIGN=top ALIGN=left>
<NOBR><A TARGET="out" HREF="out.php">^^</A
><INPUT TYPE=checkbox NAME=autoscroll <?=checked($autoscroll)?> onchange="scrollit()"
><?
$ok=0;
if ($action)
  {
    $direct="";
    $send="";
    if ($confirm=="")
      {
	$action	= htmlentities($action);
        echo "$action";
        echo " <INPUT TYPE=hidden NAME=action VALUE=\"$action\">";
        echo " <INPUT TYPE=submit NAME=confirm VALUE=yes>";
        echo " <INPUT TYPE=submit NAME=confirm VALUE=no>";
        echo "</NOBR>";
      }
    else
      {
        if ($confirm=="yes")
	  $direct	= $action;
        $action	= "";
      }
  }
else if ($shortcut)
  $direct	= $shortcut;

if ($action==""):

if ($direct!="")
  {
    $rewrite	= array(
"restart" => "exit\n",
"kill"=>"\3",
"quit"=>"\34",
"xoff"=>"\21",
"xon"=>"\23",
"eof"=>"\4",
"esc"=>"\33",
"escesc"=>"\33\33",
"(return)"=>"",
);
    $ok		= 1;
    $send	= $direct;
    if (isset($rewrite[$direct]))
      $send	= $rewrite[$direct];
  }
if ($port=="")
  $stat = "illegal sock port";
else if ($send=="" && !$ok)
  $stat	= "running $port";
else if ($fd=fsockopen("sock/$port", 0))
  {
    $stat	= "sent: ".htmlentities(addcslashes($send, "\0..\37\177..\237"));
    fwrite($fd, "$send\n");
    fflush($fd);
    flush();
    sleep(1);
    fclose($fd);
  }
else
  $stat	= "cannot open sock ".htmlentities($port);
?>
<INPUT TYPE=text SIZE=40 NAME=send VALUE="<?=htmlentities($send)?>" onkeyup=checkcr
><INPUT TYPE=submit NAME=ok VALUE=Send
><SELECT NAME=shortcut onchange="document.frm.submit()">
<OPTION VALUE=escesc>^[^[ send 2*ESC
<OPTION VALUE=esc>^[ send ESC
<OPTION VALUE=xoff>^Q send XOFF
<OPTION VALUE=xon>^S send XON
<OPTION VALUE=eof>^D send EOF
<OPTION VALUE="" SELECTED>(shortcuts)
<!-- add your favorite quick shell commands here -->
</SELECT
><SELECT NAME=action onchange="document.frm.submit()">
<OPTION VALUE=quit>^\ quit service
<OPTION VALUE=kill>^C kill service
<!-- OPTION>restart -->
<OPTION VALUE="" SELECTED>(actions)
<OPTION>exit
<!-- add additional confirmed shell commands here -->
</SELECT
><!-- INPUT TYPE=submit NAME=direct VALUE=status
--><INPUT TYPE=submit NAME=direct VALUE="(return)">
</NOBR>
</TD><TD VALIGN=top ALIGN=left BGCOLOR=#ddffdd>&nbsp;<?=htmlentities($port)?>&nbsp;
</TD><TD VALIGN=top ALIGN=left>&nbsp;<?=$stat?>
<? endif; ?>
</TD></TR></TABLE>
</FORM>
</BODY>
</HTML>
